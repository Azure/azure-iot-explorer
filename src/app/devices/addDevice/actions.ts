import actionCreatorFactory from 'typescript-fsa';
import { ADD } from '../../constants/actionTypes';
import { DEVICE } from '../../constants/actionPrefixes';
import { DeviceIdentity } from '../../api/models/deviceIdentity';

const deviceListCreator = actionCreatorFactory(DEVICE);
export const addDeviceAction = deviceListCreator.async<DeviceIdentity, DeviceIdentity>(ADD);
