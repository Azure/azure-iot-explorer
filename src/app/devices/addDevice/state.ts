/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';

export interface AddDeviceStateInterface {
    synchronizationStatus: SynchronizationStatus;
}

export const addDeviceStateInitial = Record<AddDeviceStateInterface>({
    synchronizationStatus: SynchronizationStatus.initialized
});

export type AddDeviceStateType = IM<AddDeviceStateInterface>;
