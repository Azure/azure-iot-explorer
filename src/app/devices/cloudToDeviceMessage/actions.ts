/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../constants/actionPrefixes';
import { CLOUD_TO_DEVICE_MESSAGE } from '../../constants/actionTypes';

export const cloudToDeviceMessageAction = actionCreatorFactory(DEVICECONTENT).async<CloudToDeviceMessageActionParameters, string>(CLOUD_TO_DEVICE_MESSAGE);

export interface CloudToDeviceMessageActionParameters {
    deviceId: string;
    body: string;
    properties?: Array<{key: string, value: string, isSystemProperty: boolean}>;
}
