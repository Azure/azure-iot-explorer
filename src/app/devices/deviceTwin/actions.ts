/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import * as actionPrefixes from '../../constants/actionPrefixes';
import * as actionTypes from '../../constants/actionTypes';
import { Twin } from '../../api/models/device';

const deviceContentCreator = actionCreatorFactory(actionPrefixes.DEVICECONTENT);
export const getDeviceTwinAction = deviceContentCreator.async<string, Twin>(actionTypes.GET_TWIN);
export const updateDeviceTwinAction = deviceContentCreator.async<Twin, Twin>(actionTypes.UPDATE_TWIN);
