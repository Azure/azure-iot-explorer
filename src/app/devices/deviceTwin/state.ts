import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { Twin } from '../../api/models/device';
import { SynchronizationWrapper } from '../../api/models/synchronizationWrapper';

export interface DeviceTwinStateInterface {
    deviceTwin: SynchronizationWrapper<Twin>;
}

export const deviceTwinStateInitial = Record<DeviceTwinStateInterface>({
    deviceTwin: null
});

export type DeviceTwinStateType = IM<DeviceTwinStateInterface>;
