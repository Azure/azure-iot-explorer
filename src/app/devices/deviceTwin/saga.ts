/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { all, takeLatest, takeEvery } from 'redux-saga/effects';
import { getDeviceTwinAction, updateDeviceTwinAction } from './actions';
import { getDeviceTwinSaga } from './sagas/getDeviceTwinSaga';
import { updateDeviceTwinSaga } from './sagas/updateDeviceTwinSaga';
import { loggerSaga } from '../../shared/appTelemetry/appTelemetrySaga';

export function* deviceTwinSaga() {
    yield all([
        takeLatest(getDeviceTwinAction.started.type, getDeviceTwinSaga),
        takeEvery(updateDeviceTwinAction.started.type, updateDeviceTwinSaga),
        takeEvery('*', loggerSaga)
    ]);
}
