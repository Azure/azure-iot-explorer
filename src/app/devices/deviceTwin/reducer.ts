/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { reducerWithInitialState } from 'typescript-fsa-reducers';
import { deviceTwinStateInitial, DeviceTwinStateType, DeviceTwinStateInterface } from './state';
import { getDeviceTwinAction, updateDeviceTwinAction } from './actions';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';
import { Twin } from '../../api/models/device';

export const deviceTwinReducer = reducerWithInitialState<DeviceTwinStateInterface>(deviceTwinStateInitial())
    .case(getDeviceTwinAction.started, (state: DeviceTwinStateType) => {
        return state.merge({
            deviceTwin: {
                synchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getDeviceTwinAction.done, (state: DeviceTwinStateType, payload: {params: string, result: Twin}) => {
        return state.merge({
            deviceTwin: {
                payload: payload.result,
                synchronizationStatus: SynchronizationStatus.fetched
            }
        });
    })
    .case(getDeviceTwinAction.failed, (state: DeviceTwinStateType) => {
        return state.merge({
            deviceTwin: {
                synchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(updateDeviceTwinAction.started, (state: DeviceTwinStateType) => {
        return state.merge({
            deviceTwin: {
                payload: state.deviceTwin.payload,
                synchronizationStatus: SynchronizationStatus.updating
            }
        });
    })
    .case(updateDeviceTwinAction.done, (state: DeviceTwinStateType, payload: {params: Twin, result: Twin}) => {
        return state.merge({
            deviceTwin: {
                payload: payload.result,
                synchronizationStatus: SynchronizationStatus.upserted
            }
        });
    })
    .case(updateDeviceTwinAction.failed, (state: DeviceTwinStateType) => {
        return state.merge({
            deviceTwin: {
                payload: state.deviceTwin.payload,
                synchronizationStatus: SynchronizationStatus.failed
            }
        });
    });
