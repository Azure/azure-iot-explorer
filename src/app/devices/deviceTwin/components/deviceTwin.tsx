/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString } from '../../../shared/utils/queryStringHelper';
import { getDeviceTwinAction, updateDeviceTwinAction } from '../actions';
import { REFRESH, SAVE } from '../../../constants/iconNames';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import { MultiLineShimmer } from '../../../shared/components/multiLineShimmer';
import { HeaderView } from '../../../shared/components/headerView';
import { useAsyncSagaReducer } from '../../../shared/hooks/useAsyncSagaReducer';
import { JSONEditor } from '../../../shared/components/jsonEditor';
import { deviceTwinReducer } from '../reducer';
import { deviceTwinSaga } from '../saga';
import { deviceTwinStateInitial } from '../state';
import { useBreadcrumbEntry } from '../../../navigation/hooks/useBreadcrumbEntry';
import '../../../css/_deviceTwin.scss';

export const DeviceTwin: React.FC = () => {
    const { t } = useTranslation();
    const { search } = useLocation();

    const [ localState, dispatch ] = useAsyncSagaReducer(deviceTwinReducer, deviceTwinSaga, deviceTwinStateInitial(), 'deviceTwinState');
    const twin = localState.deviceTwin && localState.deviceTwin.payload;
    const twinState = localState.deviceTwin && localState.deviceTwin.synchronizationStatus;
    const deviceId = getDeviceIdFromQueryString(search);
    const [ state, setState ] = React.useState({
        isDirty: false,
        isTwinValid: true,
        twin: JSON.stringify(twin, null, '\t')
    });

    useBreadcrumbEntry({name: 'twin'});

    React.useEffect(() => {
        dispatch(getDeviceTwinAction.started(deviceId));
    },              [deviceId]);

    const showCommandBar = () => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: t(ResourceKeys.deviceTwin.command.refresh),
                        iconProps: {iconName: REFRESH},
                        key: REFRESH,
                        name: t(ResourceKeys.deviceTwin.command.refresh),
                        onClick: handleRefresh
                    },
                    {
                        ariaLabel: t(ResourceKeys.deviceTwin.command.save),
                        disabled: !state.isDirty || !state.isTwinValid,
                        iconProps: {iconName: SAVE},
                        key: SAVE,
                        name: t(ResourceKeys.deviceTwin.command.save),
                        onClick: handleSave
                    }
                ]}
            />
        );
    };

    const handleRefresh = () => {
        dispatch(getDeviceTwinAction.started(deviceId));
    };

    const handleSave = () => {
        setState({
            ...state,
            isDirty: false,
            isTwinValid: true
        });
        dispatch(updateDeviceTwinAction.started(JSON.parse(state.twin)));
    };

    const renderTwinViewer = () => {
        if (twinState === SynchronizationStatus.working || twinState === SynchronizationStatus.updating) {
            return <MultiLineShimmer className="device-detail"/>;
        }

        return (
            <article className="device-twin device-detail">
                { twin &&
                    <JSONEditor
                        className="json-editor"
                        content={JSON.stringify(twin, null, '\t')}
                        onChange={onChange}
                    />
                }
            </article>
        );
    };

    const onChange = (data: string) => {
        let isTwinValid = true;
        try {
            JSON.parse(data);
        }
        catch  {
            isTwinValid = false;
        }
        setState({
            ...state,
            isDirty: true,
            isTwinValid,
            twin: data
        });
    };

    return (
        <>
            {showCommandBar()}
            <HeaderView
                headerText={ResourceKeys.deviceTwin.headerText}
                tooltip={ResourceKeys.deviceTwin.tooltip}
            />
            {renderTwinViewer()}
        </>
    );
};
