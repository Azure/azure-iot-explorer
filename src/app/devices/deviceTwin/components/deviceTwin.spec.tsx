/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import 'jest';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { CommandBar } from '@fluentui/react';
import { DeviceTwin } from './deviceTwin';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import * as AsyncSagaReducer from '../../../shared/hooks/useAsyncSagaReducer';
import { DeviceTwinStateInterface } from './../state';
import { getDeviceTwinAction, updateDeviceTwinAction } from '../actions';

jest.mock('react-router-dom', () => ({
    useLocation: () => ({ search: '?deviceId=test' }),
    useRouteMatch: () => ({ url: 'url', path: 'path'})
}));

// tslint:disable
const twin = {
    deviceEtag: '',
    deviceId: 'test',
    etag: 'AAAAAAAAAAE=',
    status: 'enabled',
    statusUpdateTime: '0001-01-01T00:00:00Z',
    connectionState: 'Disconnected',
    lastActivityTime: '0001-01-01T00:00:00Z',
    cloudToDeviceMessageCount: 0,
    authenticationType: 'sas',
    x509Thumbprint: {primaryThumbprint: null, secondaryThumbprint: null},
    properties: {},
    capabilities: {iotEdge: false},
    version: 1
};
// tslint:enable
const initialState: DeviceTwinStateInterface = {
    deviceTwin: {
        payload: twin,
        synchronizationStatus: SynchronizationStatus.fetched
    }
};
describe('devices/components/deviceTwin', () => {

    beforeEach(() => {
        jest.spyOn(AsyncSagaReducer, 'useAsyncSagaReducer').mockReturnValue([initialState, jest.fn()]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    context('snapshot', () => {
        it('matches snapshot', () => {
            expect(shallow(<DeviceTwin/>)).toMatchSnapshot();
        });

        it('calls refresh and save', () => {
            const realUseState = React.useState;
            const mockGetDeviceTwin = jest.spyOn(getDeviceTwinAction, 'started');
            const mockUpdateDeviceTwin = jest.spyOn(updateDeviceTwinAction, 'started');
            jest.spyOn(React, 'useState').mockImplementationOnce(() => realUseState({
                isDirty: true,
                isTwinValid: true,
                twin: JSON.stringify(twin)
            }));

            const wrapper = mount(<DeviceTwin/>);
            let commandBar = wrapper.find(CommandBar).first();
            act(() => commandBar.props().items[0].onClick(null));
            expect(mockGetDeviceTwin.mock.calls[0][0]).toEqual('test');

            commandBar = wrapper.find(CommandBar).first();
            act(() => commandBar.props().items[1].onClick(null));
            wrapper.update();
            expect(mockUpdateDeviceTwin.mock.calls[0][0]).toEqual(twin);
        });
    });
});
