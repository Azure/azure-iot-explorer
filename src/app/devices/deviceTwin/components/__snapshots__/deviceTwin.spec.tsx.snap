// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`devices/components/deviceTwin snapshot matches snapshot 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "deviceTwin.command.refresh",
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "deviceTwin.command.refresh",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "deviceTwin.command.save",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Save",
          },
          "key": "Save",
          "name": "deviceTwin.command.save",
          "onClick": [Function],
        },
      ]
    }
  />
  <HeaderView
    headerText="deviceTwin.headerText"
    tooltip="deviceTwin.tooltip"
  />
  <article
    className="device-twin device-detail"
  >
    <JSONEditor
      className="json-editor"
      content="{
	\\"deviceEtag\\": \\"\\",
	\\"deviceId\\": \\"test\\",
	\\"etag\\": \\"AAAAAAAAAAE=\\",
	\\"status\\": \\"enabled\\",
	\\"statusUpdateTime\\": \\"0001-01-01T00:00:00Z\\",
	\\"connectionState\\": \\"Disconnected\\",
	\\"lastActivityTime\\": \\"0001-01-01T00:00:00Z\\",
	\\"cloudToDeviceMessageCount\\": 0,
	\\"authenticationType\\": \\"sas\\",
	\\"x509Thumbprint\\": {
		\\"primaryThumbprint\\": null,
		\\"secondaryThumbprint\\": null
	},
	\\"properties\\": {},
	\\"capabilities\\": {
		\\"iotEdge\\": false
	},
	\\"version\\": 1
}"
      onChange={[Function]}
    />
  </article>
</Fragment>
`;
